profile/profile.go

```diff
@@ -35,69 +34,148 @@ func InitializeAllProfiles() {
+	p.TrigEventsChan = trigChan
+	p.Log = logger.ProfUeCtxLog.WithField(logger.FieldSupi, imsiStr)
+	profile.PSimUe[imsiStr] = &p
+}
 
-	if profile.PerUserTimeout == 0 {
-		profile.PerUserTimeout = profctx.PER_USER_TIMEOUT
-	}
+// option1 : Run default profile start to end..Once done Received
+//    - PROFILE_PASS
+//    - PROFILE_FAIL
+// option2 : Run custom profile start to end thorugh iterations. Once done receive
+//    - PROFILE_PASS
+//    - PROFILE_FAIL
+//    - PROFILE_HOLD optionally, if next iteration is hold
+// option3 : Pulses to come from REST Api
+//    - Start, End, Run-x, Run-y, runx/imsi1, rumy/imsi2)
+//    - Hold the calls
+//    - We should be able to pass events to profile
+
+func ExecuteProfile(profile *profctx.Profile, summaryChan chan common.InterfaceMessage) {
 
+	profile.Log.Infoln("ExecuteProfile started ")
 	var wg sync.WaitGroup
 	var Mu sync.Mutex
-	// Currently executing profile for one IMSI at a time
+
+	summary := &common.SummaryMessage{
+		ProfileType: profile.ProfileType,
+		ProfileName: profile.Name,
+		ErrorList:   make([]error, 0, 10),
+	}
+
+	defer func() {
+		summaryChan <- summary
+	}()
+
+	go func() {
+		var plock sync.Mutex
+		for {
+			select {
+			case msg := <-profile.ReadChan:
+				profile.Log.Infoln("Received trigger for profile ", msg)
+				// works only if profile is still running.
+				// Typically if execInParallel set true in profile
+				gnb, err := factory.AppConfig.Configuration.GetGNodeB(profile.GnbName)
+				if err != nil {
+					err = fmt.Errorf("Failed to fetch gNB context: %v", err)
+					summary.ErrorList = append(summary.ErrorList, err)
+					summaryChan <- summary
+					return
+				}
+
+				plock.Lock()
+				profile.UeCount = profile.UeCount + 1
+				imsi := profile.Imsi + profile.UeCount
+				imsiStr := "imsi-" + strconv.Itoa(imsi)
+				initImsi(profile, gnb, imsiStr)
+				pCtx := profile.PSimUe[imsiStr]
+				profile.Log.Infoln("pCtx ", pCtx)
+				wg.Add(1)
+				go func(pCtx *profctx.ProfileUeContext) {
+					defer wg.Done()
+					err := simue.ImsiStateMachine(profile, pCtx, imsiStr, summaryChan)
+					// Execution for the UE is complete. Count UE result as success or failure
+					Mu.Lock()
+					if err != nil {
+						summary.UeFailedCount++
+						summary.ErrorList = append(summary.ErrorList, err)
+					} else {
+						summary.UePassedCount++
+					}
+					Mu.Unlock()
+				}(pCtx)
+				plock.Unlock()
+			}
+		}
+	}()
+	imsi := profile.Imsi
 	for count := 1; count <= profile.UeCount; count++ {
 		imsiStr := "imsi-" + strconv.Itoa(imsi)
-		simUe := simuectx.NewSimUe(imsiStr, gnb, profile)
 		imsi++
-
 		wg.Add(1)
-		go func(simUe *simuectx.SimUe) {
-			defer wg.Done()
-			simue.Init(simUe)
-		}(simUe)
+		pCtx := profile.PSimUe[imsiStr]
 
-		wg.Add(1)
-		go func(simUe *simuectx.SimUe) {
+		go func(pCtx *profctx.ProfileUeContext) {
 			defer wg.Done()
-			err := ExecuteSimUe(profile, simUe, imsiStr)
+			err := simue.ImsiStateMachine(profile, pCtx, imsiStr, summaryChan)
```

Comment:
this is the function which looks similar to above

Reply:
This is for starting FSM for per UE. These UEs are configured at start